File: lib/harness_engineering/workflow.ex
defmodule HarnessEngineering.Workflow.LoadError do
  @moduledoc false

  defexception [:code, :message, :path]

  @impl true
  def exception(opts) do
    %__MODULE__{
      code: Keyword.fetch!(opts, :code),
      message: Keyword.fetch!(opts, :message),
      path: Keyword.get(opts, :path)
    }
  end
end

defmodule HarnessEngineering.Workflow.Definition do
  @moduledoc false

  defstruct config: %{}, prompt_template: "", path: nil
end

defmodule HarnessEngineering.Workflow do
  @moduledoc """
  `WORKFLOW.md` path selection and front-matter loading.
  """

  alias HarnessEngineering.Workflow.Definition
  alias HarnessEngineering.Workflow.LoadError
  alias HarnessEngineering.Workflow.Yaml

  def select_path(explicit_path, cwd \\ File.cwd!())

  def select_path(nil, cwd), do: realpath("WORKFLOW.md", cwd)

  def select_path("", cwd), do: realpath("WORKFLOW.md", cwd)

  def select_path(explicit_path, _cwd), do: realpath(explicit_path)

  def realpath(path, cwd \\ nil) do
    expanded = if cwd, do: Path.expand(path, cwd), else: Path.expand(path)
    resolve_symlinks(expanded)
  end

  defp resolve_symlinks(path, seen \\ 0)

  defp resolve_symlinks(path, seen) when seen > 32, do: path

  defp resolve_symlinks(path, seen) do
    path
    |> Path.split()
    |> resolve_parts([], seen)
  end

  defp resolve_parts([], acc, _seen), do: Path.join(acc)

  defp resolve_parts([part | rest], [], seen), do: resolve_parts(rest, [part], seen)

  defp resolve_parts([part | rest], acc, seen) do
    candidate = Path.join(acc ++ [part])

    case File.lstat(candidate) do
      {:ok, %File.Stat{type: :symlink}} ->
        resolve_link(candidate, rest, seen)

      _ ->
        resolve_parts(rest, acc ++ [part], seen)
    end
  end

  defp resolve_link(candidate, rest, seen) do
    case File.read_link(candidate) do
      {:ok, target} ->
        target_path =
          if Path.type(target) == :absolute do
            target
          else
            Path.expand(target, Path.dirname(candidate))
          end

        rest
        |> Enum.reduce(target_path, fn part, path -> Path.join(path, part) end)
        |> resolve_symlinks(seen + 1)

      {:error, _reason} ->
        Enum.reduce(rest, candidate, fn part, path -> Path.join(path, part) end)
    end
  end

  def load(path) do
    workflow_path = realpath(path)

    case File.read(workflow_path) do
      {:ok, text} ->
        parse_text(text, workflow_path)

      {:error, _reason} ->
        {:error,
         %LoadError{
           code: "missing_workflow_file",
           message: "workflow file cannot be read: #{workflow_path}",
           path: workflow_path
         }}
    end
  end

  defp parse_text(text, workflow_path) do
    if String.starts_with?(text, "---") do
      with {:ok, front_matter, body} <- split_front_matter(text, workflow_path),
           {:ok, config} <- decode_front_matter(front_matter, workflow_path) do
        {:ok,
         %Definition{
           config: config,
           prompt_template: String.trim(body),
           path: workflow_path
         }}
      end
    else
      {:ok,
       %Definition{
         config: %{},
         prompt_template: String.trim(text),
         path: workflow_path
       }}
    end
  end

  defp split_front_matter(text, path) do
    lines = String.split(text, ~r{\R}, include_captures: true)
    logical_lines = combine_line_endings(lines)

    case logical_lines do
      [first | rest] ->
        if String.trim(first) == "---" do
          split_front_matter_lines(rest, path)
        else
          {:ok, "", text}
        end

      _ ->
        {:ok, "", text}
    end
  end

  defp split_front_matter_lines(rest, path) do
    case Enum.find_index(rest, &(String.trim(&1) == "---")) do
      nil ->
        {:error,
         %LoadError{
           code: "workflow_parse_error",
           message: "workflow front matter is missing closing delimiter: #{path}",
           path: path
         }}

      index ->
        {front_matter_lines, [_delimiter | body_lines]} = Enum.split(rest, index)
        {:ok, Enum.join(front_matter_lines), Enum.join(body_lines)}
    end
  end

  defp combine_line_endings(parts), do: combine_line_endings(parts, [])

  defp combine_line_endings([], acc), do: Enum.reverse(acc)

  defp combine_line_endings([line, ending | rest], acc) when ending in ["\n", "\r\n", "\r"] do
    combine_line_endings(rest, [line <> ending | acc])
  end

  defp combine_line_endings([line | rest], acc), do: combine_line_endings(rest, [line | acc])

  defp decode_front_matter(front_matter, workflow_path) do
    case Yaml.decode(front_matter) do
      {:ok, config} when is_map(config) ->
        if Enum.all?(Map.keys(config), &is_binary/1) do
          {:ok, config}
        else
          {:error,
           %LoadError{
             code: "workflow_parse_error",
             message: "workflow front matter keys must be strings",
             path: workflow_path
           }}
        end

      {:ok, _other} ->
        {:error,
         %LoadError{
           code: "workflow_front_matter_not_a_map",
           message: "workflow front matter must decode to a map",
           path: workflow_path
         }}

      {:error, _reason} ->
        {:error,
         %LoadError{
           code: "workflow_parse_error",
           message: "workflow front matter is invalid YAML: #{workflow_path}",
           path: workflow_path
         }}
    end
  end
end

File: src/harness_engineering/workflow.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml


class WorkflowLoadError(RuntimeError):
    def __init__(self, code: str, message: str, *, path: Path | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.path = path


@dataclass(frozen=True, slots=True)
class WorkflowDefinition:
    config: dict[str, Any]
    prompt_template: str
    path: Path


def select_workflow_path(explicit_path: str | None, *, cwd: Path | None = None) -> Path:
    base = cwd or Path.cwd()
    if explicit_path:
        return Path(explicit_path).expanduser().resolve()
    return (base / "WORKFLOW.md").resolve()


def load_workflow(path: str | Path) -> WorkflowDefinition:
    workflow_path = Path(path).expanduser().resolve()
    try:
        text = workflow_path.read_text(encoding="utf-8")
    except OSError as exc:
        raise WorkflowLoadError(
            "missing_workflow_file",
            f"workflow file cannot be read: {workflow_path}",
            path=workflow_path,
        ) from exc

    config: dict[str, Any]
    body: str
    if text.startswith("---"):
        front_matter, body = _split_front_matter(text, workflow_path)
        try:
            decoded = yaml.safe_load(front_matter) if front_matter.strip() else {}
        except yaml.YAMLError as exc:
            raise WorkflowLoadError(
                "workflow_parse_error",
                f"workflow front matter is invalid YAML: {workflow_path}",
                path=workflow_path,
            ) from exc
        if decoded is None:
            decoded = {}
        if not isinstance(decoded, dict):
            raise WorkflowLoadError(
                "workflow_front_matter_not_a_map",
                "workflow front matter must decode to a map",
                path=workflow_path,
            )
        if any(not isinstance(key, str) for key in decoded):
            raise WorkflowLoadError(
                "workflow_parse_error",
                "workflow front matter keys must be strings",
                path=workflow_path,
            )
        config = decoded
    else:
        config = {}
        body = text

    return WorkflowDefinition(config=config, prompt_template=body.strip(), path=workflow_path)


def _split_front_matter(text: str, path: Path) -> tuple[str, str]:
    lines = text.splitlines(keepends=True)
    if not lines or lines[0].strip() != "---":
        return "", text
    for index, line in enumerate(lines[1:], start=1):
        if line.strip() == "---":
            return "".join(lines[1:index]), "".join(lines[index + 1 :])
    raise WorkflowLoadError(
        "workflow_parse_error",
        f"workflow front matter is missing closing delimiter: {path}",
        path=path,
    )


class WorkflowReloader:
    def __init__(self, path: str | Path) -> None:
        self.path = Path(path).expanduser().resolve()
        self.current: WorkflowDefinition | None = None
        self.last_error: WorkflowLoadError | None = None
        self._last_mtime_ns: int | None = None

    def load_initial(self) -> WorkflowDefinition:
        workflow = load_workflow(self.path)
        self.current = workflow
        self.last_error = None
        self._last_mtime_ns = self.path.stat().st_mtime_ns
        return workflow

    def reload_if_changed(self, *, force: bool = False) -> bool:
        try:
            mtime_ns = self.path.stat().st_mtime_ns
        except OSError:
            self.last_error = WorkflowLoadError(
                "missing_workflow_file",
                f"workflow file cannot be statted: {self.path}",
                path=self.path,
            )
            return False

        if not force and self._last_mtime_ns == mtime_ns:
            return False

        try:
            workflow = load_workflow(self.path)
        except WorkflowLoadError as exc:
            self.last_error = exc
            return False

        self.current = workflow
        self.last_error = None
        self._last_mtime_ns = mtime_ns
        return True
